### PR TITLE
ROS1 comms library

### PR DESCRIPTION
**Pull request by jhurliman**

### CODE REVIEW

File: packages/ros1/LICENSE [hunk 1]
```diff
@@ -0,0 +1,202 @@
+
```
Comment (defunctzombie) [inline packages/ros1/LICENSE:1]: @amacneil confirm we want to use apache v2 for this?

File: packages/ros1/sample-robot-docker/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+mkdir -p ~/catkin_ws/src
+cd ~/catkin_ws/
```
Comment (defunctzombie) [inline packages/ros1/sample-robot-docker/entrypoint.sh:4]: oh catkin - how I [haven't] missed you

File: packages/ros1/src/ITcp.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (defunctzombie) [inline packages/ros1/src/ITcp.ts:1]: File rename `Tcp.ts`? `TcpTypes.ts` or maybe just putting the bits in their own files.

File: packages/ros1/src/IXmlRpc.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (defunctzombie) [inline packages/ros1/src/IXmlRpc.ts:1]: `XmlRpcTypes` ? same thoughts as the tcp file comments

File: packages/ros1/src/RosMaster.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { URL } from "url";
+
+import { XmlRpcCreateClient, XmlRpcClient, XmlRpcResponse } from "./IXmlRpc";
+
+export default class RosMaster {
+  private _client: XmlRpcClient;
```
Comment (defunctzombie) [inline packages/ros1/src/RosMaster.ts:10]: lets start using the new `#` syntax for private fields (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields). This ensures true private behavior. You don't need the `private` keyword with them.

File: packages/ros1/src/RosMaster.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+  constructor(options: {
+    xmlRpcCreateClient: XmlRpcCreateClient;
+    uri?: URL;
```
Comment (defunctzombie) [inline packages/ros1/src/RosMaster.ts:14]: I'd consider only supporting uri (via a string) so the user should always pass in something of the form of protocol://host:port and we decode ourselves - feels like fewer code paths and less variance at the use site.

File: packages/ros1/src/IXmlRpc.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+export type XmlRpcValue = string | number | XmlRpcValue[];
+
+export type XmlRpcResponse = [code: number, msg: string, value: XmlRpcValue];
+
+export interface XmlRpcClient {
+  methodCall(method: string, args: XmlRpcValue[]): Promise<XmlRpcResponse>;
+}
+
+export interface XmlRpcCreateClient {
```
Comment (defunctzombie) [inline packages/ros1/src/IXmlRpc.ts:13]: Conventionally I've seen interfaces like this called `XmlRpcClientConstructor` and you should add a `new` in front of the function.

File: packages/ros1/src/RosMaster.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+import { URL } from "url";
+
+import { XmlRpcCreateClient, XmlRpcClient, XmlRpcResponse } from "./IXmlRpc";
+
+export default class RosMaster {
+  private _client: XmlRpcClient;
+
+  constructor(options: {
+    xmlRpcCreateClient: XmlRpcCreateClient;
```
Comment (defunctzombie) [inline packages/ros1/src/RosMaster.ts:13]: I think this should accept type `XmlRpc` rather than a create function and leave creation to the user

File: packages/ros1/src/node/xmlrpc-rosnodejs.d.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (defunctzombie) [inline packages/ros1/src/node/xmlrpc-rosnodejs.d.ts:1]: Custom typings customarily live in a `typings` folder - see how we've set this up for the root project.

File: packages/ros1/README.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# node-ros1
+
+node.js TypeScript library for interfacing with ROS1.
```
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/ros1/README.md:3]: You're right. This started as a node library and became more generic as I went.

File: packages/ros1/LICENSE (hunk 1 above)
Comment (amacneil) [reply packages/ros1/LICENSE:1]: can we keep everything in monorepo MPL for now and relicense if/when we separate it? I would like to have a clear licensing story rather than files everywhere with different licenses

Comment (amacneil) [reply packages/ros1/LICENSE:1]: either that or just start with a separate repo on day 1, but that sounds painful to iterate on to me.

also I think MIT is fine (rather than apache) for anything we decide not to use MPL for, it's a bit simpler / less overhead

File: packages/ros1-node/README.md [hunk 11]
```diff
@@ -0,0 +1 @@
+# @foxglove/ros1-node
```
Comment (defunctzombie) [inline packages/ros1-node/README.md:1]: consider `ros1-nodejs` just cause "node" is already such an overloaded term in rosland

File: packages/ros1/src/IXmlRpc.ts (hunk 7 above)
Comment (jhurliman) [reply packages/ros1/src/IXmlRpc.ts]: Add a `new` where? When calling it? What happens if you don't?

Comment (defunctzombie) [reply packages/ros1/src/IXmlRpc.ts]:
```typescript
export interface XmlRpcConstructor {
   new (...): XmlRpc
}
```

This is an interface that describes a constructor. From what I saw in your stuff tho I would accept XmlRpc rather than building it - let the user build it.

File: packages/ros1/src/RosSlave.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    if (Array.isArray(proto) && proto.length > 0) {
+      if (proto[0] === "TCPROS") {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+export class RosSlave {
```
Comment (jtbandes) [inline packages/ros1/src/RosSlave.ts:35]: I know some of these things reference core ROS concepts (e.g. "ros master"), but I think it is worth the effort now to at least replace 'slave' with something like 'follower', 'client' etc. See some examples at https://en.wikipedia.org/wiki/Master/slave_(technology)#Terminology_concerns

File: packages/ros1-nodejs/README.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# @foxglove/ros1-nodejs
+
+node.js implementation for the @foxglove/ros1 library.
+
+### Usage
+
+TODO
```
Comment (defunctzombie) [inline packages/ros1-nodejs/README.md:7]: Remove `Usage` section - we can revisit later. TODO looks lame if someone opens the readme.

File: packages/ros1-nodejs/src/PlatformNode.ts [hunk 14]
```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import os from "os";
+import { URL } from "whatwg-url";
+
+function IsPrivateIP(ip: string): boolean {
```
Comment (jtbandes) [inline packages/ros1-nodejs/src/PlatformNode.ts:8]: naming: capitals should be reserved for type/class/component names, lowerCamel for functions
```suggestion
function isPrivateIP(ip: string): boolean {
```

File: packages/ros1-nodejs/package.json [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  "private": true,
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/foxglove/studio.git"
+  },
+  "author": {
+    "name": "Foxglove Technologies",
+    "email": "support@foxglove.dev"
+  },
+  "homepage": "https://github.com/foxglove/studio",
```
Comment (defunctzombie) [inline packages/ros1-nodejs/package.json:15]: https://foxglove.dev

File: packages/ros1-nodejs/package.json [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  "main": "src/index.ts",
+  "module": "src/index.ts",
+  "files": [
+    "src"
+  ],
+  "engines": {
+    "node": ">=12"
```
Comment (defunctzombie) [inline packages/ros1-nodejs/package.json:22]: Maybe omit engines field? Top level package.json can specify it if we care.

File: packages/ros1-nodejs/package.json [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  "scripts": {},
+  "prettier": {
```
Comment (defunctzombie) [inline packages/ros1-nodejs/package.json:25]: Remove - we should use the prettier config in the top level of the repo to keep one consistent config.

File: packages/ros1-nodejs/src/PlatformNode.ts [hunk 18]
```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/PlatformNode.ts:1]: Not sure what ties all the functions in this file together. Maybe a comment indicating what should be in this file so it doesn't become a dumping ground for _stuff_.

File: packages/ros1-nodejs/src/TcpMessageStream.ts [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { Transform, TransformCallback } from "stream";
+
+export class TcpMessageStream extends Transform {
+  #inMessage = false;
+  #bytesNeeded = 4;
+  #chunks: Uint8Array[] = [];
+
+  _transform(chunk: Uint8Array, _encoding: BufferEncoding, callback: TransformCallback): void {
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpMessageStream.ts:12]: Prefer `private` typescript keyword and no `_` prefix.

File: packages/ros1-nodejs/src/TcpMessageStream.ts [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+      }
+
+      this.#inMessage = !this.#inMessage;
+      this.#chunks = [];
+    }
+
+    callback();
+  }
+
+  static ConcatData(chunks: Uint8Array[]): Uint8Array {
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpMessageStream.ts:47]: Looks like this is unrelated to this class and belongs in a separate file.

File: packages/ros1-nodejs/src/TcpNode.ts [hunk 21]
```diff
@@ -0,0 +1,149 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { TcpAddress, TcpServer, TcpSocket } from "@foxglove/ros1";
+import EventEmitter from "eventemitter3";
+import * as net from "net";
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpNode.ts:7]: `import net from "net";` should work, no?

File: packages/ros1-nodejs/src/TcpNode.ts [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+    const family = this.#socket.remoteFamily; // There is no localFamily
+    const address = this.#socket.localAddress;
+    return port !== undefined && family !== undefined && address !== undefined
+      ? { port, family, address }
+      : undefined;
+  }
+
+  fd(): number | undefined {
+    // eslint-disable-next-line no-underscore-dangle
+    return ((this.#socket as unknown) as MaybeHasFd)._handle?.fd;
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpNode.ts:63]: What stopped the cast from this.#socket directly to `MaybeHasFd` ? Do sockets not always have `_handle`? Is there a way to specify that `this.#socket` should be a socket with a handle or we don't want that limitation?

File: packages/ros1-nodejs/src/TcpNode.ts [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+          reject(err);
+          return;
+        }
+        resolve();
+      });
+    });
+  }
+}
+
+export class TcpServerNode extends EventEmitter implements TcpServer {
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpNode.ts:87]: Lets try to do one class per file. So `TcpServerNode.ts`, `TcpNode.ts`.

File: packages/ros1-nodejs/src/TcpNode.ts [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    return addr;
+  }
+
+  close(): void {
+    this.#server.close();
+  }
+}
+
+export function TcpListen(options: {
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpNode.ts:114]: Separate file.

File: packages/ros1-nodejs/src/TcpNode.ts [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+    const server = net.createServer();
+    server.on("error", reject);
+    server.listen(options.port, options.host, options.backlog, () => {
+      server.removeListener("error", reject);
+      resolve(new TcpServerNode(server));
+    });
+  });
+}
+
+export function TcpConnect(options: { host: string; port: number }): Promise<TcpSocket> {
```
Comment (defunctzombie) [inline packages/ros1-nodejs/src/TcpNode.ts:129]: Separate file

File: packages/ros1-nodejs/tsconfig.json [hunk 26]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (defunctzombie) [inline packages/ros1-nodejs/tsconfig.json:1]: This file should extend our base tsconfig.json in repo root and override only a minimal set of this required for this project.

File: packages/ros1-turtlesim-test/tsconfig.json [hunk 27]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (defunctzombie) [inline packages/ros1-turtlesim-test/tsconfig.json:1]: This file should extend our base tsconfig.json

File: packages/ros1/package.json [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+  "main": "src/index.ts",
+  "module": "src/index.ts",
+  "files": [
+    "src"
+  ],
+  "engines": {
+    "node": ">=12"
+  },
+  "scripts": {},
+  "prettier": {
```
Comment (defunctzombie) [inline packages/ros1/package.json:25]: use prettier from our top level config - should not have a separate one

File: packages/ros1/src/TcpConnection.ts [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+import { MessageReader, parseMessageDefinition, RosMsgDefinition } from "rosbag";
+import { TextDecoder, TextEncoder } from "web-encoding";
+
+import { Connection } from "./Connection";
+import { TcpAddress, TcpSocket } from "./TcpTypes";
+
+type Header = [string, string][];
+
+export class TcpConnection implements Connection {
+  _socket: TcpSocket;
```
Comment (defunctzombie) [inline packages/ros1/src/TcpConnection.ts:14]: Prefer `#` for private fields.

File: packages/ros1/src/TcpConnection.ts [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+    const addr = this._socket.remoteAddress();
+    const fd = this._socket.fd() ?? -1;
+    if (addr) {
+      const { address, port } = addr;
+      return `TCPROS connection on port ${localPort} to [${address}:${port} on socket ${fd}]`;
+    }
+    return `TCPROS not connected [socket ${fd}]`;
+  }
+
+  #handleClose = (): void => {
```
Comment (defunctzombie) [inline packages/ros1/src/TcpConnection.ts:61]: Were you trying different approaches to making private methods? Should we be doing this or using `private` keyword for methods? Ideally we are consistent

File: packages/ros1/src/index.ts [hunk 31]
```diff
@@ -0,0 +1,23 @@
...
+export * from "./RosSlave";
+export * from "./RosSlaveClient";
+export * from "./SubscriberLink";
+export * from "./Subscription";
+export * from "./TcpConnection";
+export * from "./TcpTypes";
+export * from "./XmlRpcTypes";
+
+// Mock backend
+export * as PlatformMock from "./mock/PlatformMock";
```
Comment (defunctzombie) [inline packages/ros1/src/index.ts:21]: Think it is useful to export the mocks?

File: packages/ros1/src/mock/TcpMock.ts [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+  address(): TcpAddress | undefined {
+    return this.listening ? { address: "192.168.1.1", port: 20000, family: "IPv4" } : undefined;
+  }
+
+  close(): void {
+    this.listening = false;
+  }
+}
+
+export function TcpListen(_options: {
```
Comment (defunctzombie) [inline packages/ros1/src/mock/TcpMock.ts:57]: I think you can omit the options arg altogether? Or does that fail typechecks when used in other places?

File: packages/ros1-nodejs/src/TcpNode.ts [hunk 33]
```diff
@@ -0,0 +1,149 @@
...
+
+type MaybeHasFd = {
+  _handle:
+    | {
+        fd: number | undefined;
+      }
+    | undefined;
+};
+
+export class TcpSocketNode extends EventEmitter implements TcpSocket {
```
Comment (jtbandes) [inline packages/ros1-nodejs/src/TcpNode.ts:19]: Does this "node" mean "node.js"? Maybe call it `TcpSocketNodeJS` to be clearer.

File: packages/ros1/tsconfig.json [hunk 34]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (defunctzombie) [inline packages/ros1/tsconfig.json:1]: Should extend the tsconfig.json at project root

File: packages/ros1/src/RosSlave.ts [hunk 35]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    const publications = this.#rosNode.publications.values();
+    const subscriptions = this.#rosNode.subscriptions.values();
+
+    const publishStats: XmlRpcValue[] = Array.from(publications, (pub, _) => pub.getStats());
+    const subscribeStats: XmlRpcValue[] = Array.from(subscriptions, (sub, _) => sub.getStats());
+    const serviceStats: XmlRpcValue[] = [];
+
+    return Promise.resolve([1, "test", [publishStats, subscribeStats, serviceStats]]);
```
Comment (jtbandes) [inline packages/ros1/src/RosSlave.ts:84]: is this "test" a TODO?

File: packages/ros1/src/RosNode.ts [hunk 36]
```diff
@@ -0,0 +1,186 @@
...
+    // Register with each publisher
+    await Promise.all(
+      publishers.map(async (pubUrlStr) => {
+        const url = ToUrl(pubUrlStr as string);
+        if (url === undefined) {
+          return;
+        }
+
+        // Create an XMLRPC client to talk to this publisher
+        const xmlRpcClient = await this.#xmlRpcCreateClient({ url });
```
Comment (jtbandes) [inline packages/ros1/src/RosNode.ts:136]: why does creation need to be an async call?

I wonder if it would be cleaner for RosSlaveClient to make the xmlrpc client itself, just given the url.

File: packages/ros1/src/TcpConnection.ts [hunk 37]
```diff
@@ -0,0 +1,133 @@
...
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { MessageReader, parseMessageDefinition, RosMsgDefinition } from "rosbag";
+import { TextDecoder, TextEncoder } from "web-encoding";
+
+import { Connection } from "./Connection";
+import { TcpAddress, TcpSocket } from "./TcpTypes";
+
+type Header = [string, string][];
```
Comment (jtbandes) [inline packages/ros1/src/TcpConnection.ts:11]: I think this would be nicer as an object/record, unless duplicate keys are needed? I also see you are using a Map elsewhere.

File: packages/ros1/src/TcpConnection.ts [hunk 38]
```diff
@@ -0,0 +1,133 @@
...
+  static ParseHeader(data: Uint8Array): Map<string, string> {
+    const decoder = new TextDecoder();
+    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
+    const result = new Map<string, string>();
+
+    let idx = 0;
+    while (idx + 4 < data.length) {
+      const len = Math.min(view.getUint32(idx, true), data.length - idx - 4);
+      idx += 4;
+      const str = decoder.decode(new Uint8Array(data.buffer, data.byteOffset + idx, len)) as string;
```
Comment (jtbandes) [inline packages/ros1/src/TcpConnection.ts:120]: why is `as string` needed here?

File: packages/ros1/src/TcpConnection.ts [hunk 39]
```diff
@@ -0,0 +1,133 @@
...
+      let equalIdx = str.indexOf("=");
+      if (equalIdx < 0) {
+        equalIdx = str.length;
+      }
+      const key = str.substr(0, equalIdx);
+      const value = str.substr(equalIdx + 1);
+      result.set(key, value);
+      idx += len;
+    }
```
Comment (jtbandes) [inline packages/ros1/src/TcpConnection.ts:129]: warn if leftover bytes?

File: packages/ros1/src/RosSlave.ts (hunk 12 above)
Comment (jhurliman) [reply packages/ros1/src/RosSlave.ts:35]: I struggled with this. I initially had a different set of terminology and was getting lost in my translation from ROS concepts to my own. Now that the implementation is fleshed out and working I think renaming is more feasible. I'll do a pass now.

File: packages/ros1-nodejs/src/TcpMessageStream.ts (hunk 19 above)
Comment (jhurliman) [reply packages/ros1-nodejs/src/TcpMessageStream.ts:12]: It's not my API :woman_shrugging:

File: packages/ros1-nodejs/src/TcpNode.ts (hunk 22 above)
Comment (jhurliman) [reply packages/ros1-nodejs/src/TcpNode.ts:63]: Sockets do not always have a handle, only one some platforms. It's also only documented through folklore and reading the internals of node.js source code, but this is the only way (in pure JS at least) to access the fd for a socket.

File: packages/ros1-nodejs/src/TcpMessageStream.ts (hunk 19 above)
Comment (defunctzombie) [reply packages/ros1-nodejs/src/TcpMessageStream.ts:12]: ah yes I see that - happy times

File: packages/ros1/src/mock/TcpMock.ts (hunk 32 above)
Comment (jhurliman) [reply packages/ros1/src/mock/TcpMock.ts:57]: It no longer matches the interface definition so typechecking fails when used

File: packages/ros1-nodejs/src/PlatformNode.ts [hunk 40]
```diff
@@ -0,0 +1,80 @@
...
+export function GetDefaultRosMasterUri(): Promise<URL> {
+  const urlStr = process.env["ROS_MASTER_URI"];
+  if (urlStr !== undefined && urlStr.length > 0) {
+    try {
+      const url = new URL(urlStr);
+      if (url.protocol === "http:" || url.protocol === "https:") {
+        return Promise.resolve(url);
+      }
+    } finally {
+      // Ignore malformed ROS_MASTER_URI env var
```
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/ros1-nodejs/src/PlatformNode.ts:26]: There are a handful of places in these libraries where dependency injecting a logger could be useful. I can create a followup ticket for that.

File: packages/ros1-nodejs/src/TcpNode.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/ros1-nodejs/src/TcpNode.ts:63]: `_handle?` doesn't assume `_handle` exists. This is the only way of accessing the file descriptor for a socket from pure JS, on platforms where sockets have file descriptors. This message isn't logged, it is sent over the network to ROS code which may be making terrible assumptions about the formatting of this string and `fd` being properly set on Linux. I'm trying to implement the ROS protocol as closely as possible to roscpp.

[the remark this answers is not in the document]
Comment (jtbandes) [reply packages/ros1-nodejs/src/TcpNode.ts:63]: Is that even possible? What could someone on another host meaningfully do with information about which fd someone else's machine is using for their socket?

File: packages/ros1/src/RosNode.ts (hunk 36 above)
Comment (jhurliman) [reply packages/ros1/src/RosNode.ts:136]: Because the `node-electron` implementation will look like "send a message to a separate process requesting this thing be made, instantiate it in that separate process, return a message saying it has been instantiated with a handle referencing it"

File: packages/ros1/src/TcpConnection.ts (hunk 37 above)
Comment (jhurliman) [reply packages/ros1/src/TcpConnection.ts:11]: It would be nicer but less powerful, since you would not be able to control what the final serialization looked like. I did it this way to exactly match the roscpp serialization.

File: packages/ros1/src/TcpConnection.ts (hunk 38 above)
Comment (jhurliman) [reply packages/ros1/src/TcpConnection.ts:120]: Because `TextDecoder` is type `any` :(.

File: packages/ros1-nodejs/src/TcpNode.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/ros1-nodejs/src/TcpNode.ts]: It's a little unclear what the purpose of this data is at all, but I'm guessing it would be fetched and stored for some sort of logging system. Why the local socket fd matters, especially when that concept doesn't even exist on all platforms, is beyond me.

File: packages/ros1/src/RosSlave.ts (hunk 12 above)
Comment (amacneil) [reply packages/ros1/src/RosSlave.ts]: Also it seems like ros themselves are moving more towards names like "roscore" over "rosmaster", although maybe they haven't fully committed to it because both seem to exist?

File: packages/ros1-nodejs/src/TcpNode.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (jtbandes) [reply packages/ros1-nodejs/src/TcpNode.ts]: So can we just remove this for now since we don't think it's being used for any important purpose? Would be nice to clean up the strange TS hackery unless we know it's valuable.
